indexer/jobs/base_job.py

```diff
@@ -12,6 +14,42 @@
 from indexer.domain.transaction import Transaction
 from indexer.utils.reorg import should_reorg
 
+T = TypeVar("T")
+
+
+class Collector(Generic[T]):
+
+    def __init__(self, job, collect_types: List[Domain]):
+        self.job = job
+        self.collect_types = set(collect_types)
+
+    def check_collect_type(self, cls):
+        if cls not in self.collect_types:
+            raise ValueError(
+                f"Collector only collects the following types of data: {[self.collect_types]}, "
+                f"the data type given now is {cls}"
+            )
+
+    def collect_item(self, key: str, data: Domain):
+        self.check_collect_type(type(data))
+        with self.job._data_buff_lock[key]:
+            self.job._data_buff[key].append(data)
+
+    def collect_items(self, key, datas: List[Domain]):
+        self.check_collect_type(type(datas[0]))
+        with self.job._data_buff_lock[key]:
+            self.job._data_buff[key].extend(datas)
+
+    def collect(self, domain: Domain):
```

Comment:
Let's keep collect_domain/collect_domains 

Or you can add both collect/collect_domain